Date: Wed, 6 Aug 2014 13:20:41 +0200
From: Libor Pechacek <lpechacek@suse.cz>
Subject: [PATCH] kgr: mark kernel unsupported upon patch revert
Patch-mainline: never, SLE specific
References: fate#313296

Reverting kGraft patches in not yet supported in SUSE Linux Enterprise.

js: move to a separate function

Signed-off-by: Libor Pechacek <lpechacek@suse.cz>
Acked-by: Jiri Kosina <jkosina@suse.cz>
Signed-off-by: Jiri Slaby <jslaby@suse.cz>
---
 kernel/kgraft_files.c |   18 ++++++++++++++++++
 1 file changed, 18 insertions(+)

--- a/kernel/kgraft_files.c
+++ b/kernel/kgraft_files.c
@@ -75,12 +75,30 @@ static ssize_t replace_all_show(struct k
 	return snprintf(buf, PAGE_SIZE, "%d\n", p->replace_all);
 }
 
+static void kgr_taint_kernel(const struct kgr_patch *p)
+{
+#ifdef CONFIG_SUSE_KERNEL_SUPPORTED
+	const char *modname;
+
+#ifdef CONFIG_MODULES
+	modname = p->owner ? p->owner->name : "n/a";
+#else
+	modname = "n/a";
+#endif
+	pr_warning("attempt to revert kgr patch %s (%s), setting NO_SUPPORT taint flag\n",
+			p->name, modname);
+	add_taint(TAINT_NO_SUPPORT, LOCKDEP_STILL_OK);
+#endif
+}
+
 static ssize_t revert_store(struct kobject *kobj,
 		struct kobj_attribute *attr, const char *buf, size_t count)
 {
 	struct kgr_patch *p = kobj_to_patch(kobj);
 	int ret;
 
+	kgr_taint_kernel(p);
+
 	ret = kgr_modify_kernel(p, true);
 
 	return ret < 0 ? ret : count;
